pipeline/backend/lxc/lxc.go

```diff
@@ -0,0 +1,232 @@
+// Copyright 2023 Woodpecker Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lxc
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"regexp"
+	"strings"
+	"text/template"
+
+	"github.com/rs/zerolog/log"
+
+	"github.com/woodpecker-ci/woodpecker/pipeline/backend/types"
+	"github.com/woodpecker-ci/woodpecker/shared/constant"
+)
+
+var scriptTemplate = template.Must(template.New("script").Parse(`#!/bin/sh -xe
+lxc-create --name="{{.Name}}" --template={{.Template}} -- --release {{.Release}} --package=dbus
+tee -a /var/lib/lxc/{{.Name}}/config <<'EOF'                                                                                                                   
+security.nesting = true
+lxc.cap.drop =
+lxc.apparmor.profile = unconfined
```

Comment:
Yes, it is. 

This backend, very much like the local backend, requires a high level of trust. The primary motivation is not secure confinement of the jobs. It is to provide a way to create a fresh systemd capable environment virtually identical to a real machine that can be wiped out and leave no trace behind when it completes. It guards against breakage caused by failed tests but not against malicious jobs.

Maybe it is possible to make this LXC backend very secure but I'm not aware that containers that provide root access can do that effectively.

Reply:
You could run rootless containers, running LXC as a user and not as root is a first step. And you could probably create a new user in the LXC container, so it's not root.
But LXC could be described as a glorified chroot. So even though it gives some flexibility, it doesn't provide much security.

Is there a specific reason for disabling apparmor ?

Reply:
Well put. There is no specific reason to disable apparmor, only it makes it easier because I do not fully understand how to properly harden the underlying LXC container running on the host in a way that would allow nested LXC containers to get all the capabilities they want (including everything down to disabling any protection, but only for nested containers).

Reply:
Disabling Apparmor could also be restricted by a variable `AGENT_LXC_DISABLE_APPARMOR`. That way it's optional.

But after some thinking, is it possible to run this with rootless containers ? Do you mount any filsystems in the container ?

Reply:
Because it cannot be configured [to allow essential operations such as mounting file systems](https://discuss.linuxcontainers.org/t/what-does-security-nesting-true/7156/2). It would be possible to keep it activated and run unprivileged containers. But that would offer no benefit compared to docker containers.